### PR TITLE
Encoding Max Type Variables in manifest2proto

### PR DESCRIPTION
**Pull request by alxrsngrtn**
Adding a flag to indicate if a type variable is was resolved to max type. 

CC @bgogul

### CODE REVIEW

File: java/arcs/core/data/proto/manifest.proto [hunk 1]
```diff
@@ -156,6 +156,7 @@ message ReferenceTypeProto {
 message TypeVariableProto {
   string name = 1;
   ConstraintInfo constraint = 2;
+  bool maxType = 3;
```
Comment (piotrswigon) [inline java/arcs/core/data/proto/manifest.proto:159]: I think we should rename this to something like `accessAllFields`.

The reason is that the feature that `with {*}` represents is being able to access all fields from the resolved type. Making the type system choose the max type follows from that, but is not primary. E.g. If it was only the type system choosing the max type then DFA should not behave differently - but it's not about that - it's about what the particle can access, so DFA needs to be impacted.

If you argue that we should be consistent with the TypeScript side, then I would say that this ship has sailed already. In the protocol buffer we use type variables exclusively to signify generics in connections, which is why we decided to not have canReadSubset and canWriteSuperset. Instead we just have one constraint that represents fields that the particle will read. 

Maybe @bgogul can comment and convince me otherwise.

File: src/tools/tests/manifest2proto-test.ts [hunk 2]
```diff
@@ -712,6 +714,56 @@ describe('manifest2proto', () => {
...
 
+  it('encodes max variable type - writeSuperset constraint', async () => {
+    const constraint = EntityType.make(['Foo'], {value: 'Text'}).singletonOf();
+    const varType = TypeVariable.make('a', constraint, null, true);
+    assert.deepStrictEqual(await toProtoAndBackType(varType), {
+      variable: {
+        name: 'a',
+        maxType: true,
+        constraint: {constraintType: {
+            singleton: {singletonType: {
```
Comment (piotrswigon) [inline src/tools/tests/manifest2proto-test.ts:725]: This indentation is weird. Either do it one element per line, e.g.

```
a: {
  b: {
    c: 'foo'
  }
}
```

or inline sometimes while ensuring that the next indent is +2 spaces:

```
a: {b: {
  c: 'foo'
}}
```

But don't mix it in a way that makes following indent of arbitrary depth.

**Comment by piotrswigon** [comment]
Nice quick turn around BTW!

File: src/tools/tests/manifest2proto-test.ts (hunk 2 above)
Comment (alxrsngrtn) [reply src/tools/tests/manifest2proto-test.ts:725]: This style was adapted after a suggestion from Myk, see: https://github.com/PolymerLabs/arcs/pull/5302#discussion_r425501978

Is there a middle ground that we can settle on?

Comment (alxrsngrtn) [reply src/tools/tests/manifest2proto-test.ts:725]: This indentation is also consistent w/ the rest of the tests in this file (and others).

Comment (piotrswigon) [reply src/tools/tests/manifest2proto-test.ts:725]: In Myk's suggestion every following level of indentation is 2 spaces, I am happy with more compactness, but I don't like this:
```
a: {b: {c: {d: {
        x: 'foo'
      }
    }
  }
}
```

Comment (alxrsngrtn) [reply src/tools/tests/manifest2proto-test.ts:725]: Ah, I see what you mean about mixing. Cool, it's clear that this should change and how it should change.

Comment (alxrsngrtn) [reply src/tools/tests/manifest2proto-test.ts:725]: Huh, the view in my IDE is different.

Comment (alxrsngrtn) [reply src/tools/tests/manifest2proto-test.ts:725]: No, the file is just really inconsistent -- I'll take the liberty of cleaning it up, generally.

File: java/arcs/core/data/proto/manifest.proto (hunk 1 above)
Comment (bgogul) [reply java/arcs/core/data/proto/manifest.proto:159]: I agree `maxType` is not quite intuitive. Also, this new field should go into the `TypeConstraintsProto` field as this essentially part of the constraints associated with the type variable? A couple of suggestions:
  - Add this field to `TypeConstraintsProto`.
  - Name it as `unconstrained` or `no_constraints`? (Alternatively, we can name it as `has_constraints`, which should be set to `false` if `maxType` is `true` and vice versa.)
  - In the Kotlin data class for `TypeVariable`, add a property `noConstraints` or `unconstrained` that is set appropriately. (Ideally, we want Kotlin to have a `TypeConstraints` class, but let us defer it for later as we will need to make some change in google3 simultaneously for that.)

Comment (alxrsngrtn) [reply java/arcs/core/data/proto/manifest.proto]: I was resistant to the term "unconstrained" on first read, but after looking at the code, I see its appeal. Should be implemented now.

File: src/tools/manifest2proto.ts [hunk 3]
```diff
@@ -336,7 +336,10 @@ export async function typeToProtoPayload(type: Type) {
     case 'TypeVariable': {
       const constraintType = type.canReadSubset || type.canWriteSuperset;
       const name = {name: (type as TypeVariable).variable.name};
-      const constraint = constraintType ? {constraint: {constraintType: await typeToProtoPayload(constraintType)}} : {};
+      const constraint = {constraint: {unconstrained: (type as TypeVariable).variable.resolveToMaxType || false}};
+      if (constraintType) {
+        constraint.constraint['constraintType'] = await typeToProtoPayload(constraintType);
```
Comment (bgogul) [inline src/tools/manifest2proto.ts:341]: do we need to add a check that if `resolveToMaxType` is `true`, `constraintType` should be null?

File: src/tools/tests/manifest2proto-test.ts [hunk 4]
```diff
@@ -692,21 +642,55 @@ describe('manifest2proto', () => {
...
+  });
+
+  it('encodes max variable type - readSubset constraint', async () => {
+    const constraint = EntityType.make(['Foo'], {value: 'Text'}).singletonOf();
+    const varType = TypeVariable.make('a', null, constraint, true);
+    assert.deepStrictEqual(await toProtoAndBackType(varType), {
+      variable: {
+        name: 'a',
+        constraint: {
+          constraintType: {singleton: {singletonType: {
```
Comment (bgogul) [inline src/tools/tests/manifest2proto-test.ts:686]: Shouldn't constraintType be empty if `unconstrained` is true.

File: java/arcs/core/data/proto/manifest.proto [hunk 5]
```diff
@@ -166,6 +166,7 @@ message ListTypeProto {
 // not those resulting from type inference.
 message ConstraintInfo {
   TypeProto constraint_type = 1;
+  bool unconstrained = 2;
```
Comment (bgogul) [inline java/arcs/core/data/proto/manifest.proto:169]: When reading the code, `no_constraints` seems slightly more readable. e.g., `no_constraints` is `true` has some implication on the `constraint_type`.

(Sorry for the back and forth. What do you think?)

File: src/tools/tests/manifest2proto-test.ts [hunk 6]
```diff
@@ -777,7 +786,25 @@ describe('manifest2proto', () => {
 
     assert.deepStrictEqual(varInput, varOutput);
     assert.deepStrictEqual(varInput.name, 'a');
-    assert.isUndefined(varInput.constraint);
+    assert.isFalse(varInput.constraint.unconstrained);
+    assert.isFalse(varOutput.constraint.unconstrained);
+  });
+
+  it('encodes max variable type for particle specs - unconstrained', async () => {
```
Comment (bgogul) [inline src/tools/tests/manifest2proto-test.ts:793]: Can we add the following test where `unconstrained` should be false and `constraintType` is empty (or null).

```
Particle P
   input: reads ~a 
   output: writes ~a
```

File: src/tools/tests/manifest2proto-test.ts [hunk 7]
```diff
@@ -758,9 +743,33 @@ describe('manifest2proto', () => {
...
+    `);
+
+    const particleSpec = (await toProtoAndBack(manifest)).particleSpecs[0];
+    const varInput = particleSpec.connections.find(c => c.name === 'input').type.variable;
+    const varOutput = particleSpec.connections.find(c => c.name === 'output').type.variable;
+
+    assert.deepStrictEqual(varInput, varOutput);
+    assert.deepStrictEqual(varInput.name, 'a');
+    assert.deepStrictEqual(varInput.constraint, {
+      constraintType: {
```
Comment (bgogul) [inline src/tools/tests/manifest2proto-test.ts:766]: Shoudn't this be empty?

**Comment by alxrsngrtn** [comment]
Closing in favor of #5893, it includes these changes.
